contract_payment_mode/tests/test_contract_payment.py

```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+# Copyright 2015 Antiun Ingenieria S.L. - Antonio Espinosa
+# Copyright 2017 Tecnativa - Vicent Cubells
+# Copyright 2017 Tecnativa - David Vidal
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo.tests import common
+from ..hooks import post_init_hook
+
+
+class TestContractPaymentInit(common.SavepointCase):
+    @classmethod
+    def setUpClass(cls):
+        super(TestContractPaymentInit, cls).setUpClass()
+
+        cls.payment_method = cls.env['account.payment.method'].create({
+            'name': 'Test Payment Method',
+            'code': 'Test',
+            'payment_type': 'inbound',
+        })
+        cls.payment_mode = cls.env['account.payment.mode'].create({
+            'name': 'Test payment mode',
+            'active': True,
+            'payment_method_id': cls.payment_method.id,
+            'bank_account_link': 'variable',
+        })
+        cls.partner = cls.env['res.partner'].create({
+            'name': 'Test contract partner',
+            'customer_payment_mode_id': cls.payment_mode,
+        })
+        cls.product = cls.env['product.product'].create({
+            'name': 'Custom Service',
+            'type': 'service',
+            'uom_id': cls.env.ref('product.product_uom_hour').id,
+            'uom_po_id': cls.env.ref('product.product_uom_hour').id,
+            'sale_ok': True,
+        })
+        cls.contract = cls.env['account.analytic.account'].create({
+            'name': 'Maintenance of Servers',
+        })
+
+    def test_post_init_hook(self):
+        contract = self.env['account.analytic.account'].create({
+            'name': 'Test contract',
+            'partner_id': self.partner.id,
+            'payment_mode_id': self.payment_mode.id,
+        })
+        self.assertEqual(contract.payment_mode_id,
+                         self.payment_mode)
+
+        contract.payment_mode_id = False
+        self.assertEqual(contract.payment_mode_id.id, False)
+
+        post_init_hook(self.cr, self.env)
+        self.assertEqual(contract.payment_mode_id,
+                         self.payment_mode)
+
+    def test_contract_and_invoices(self):
+        self.contract.write({'partner_id': self.partner.id})
+        self.contract.on_change_partner_id()
+        self.assertEqual(self.contract.payment_mode_id,
+                         self.contract.partner_id.customer_payment_mode_id)
+        self.contract.write({
+            'recurring_invoices': True,
+            'recurring_interval': 1,
+            'recurring_invoice_line_ids': [(0, 0, {
+                'quantity': 2.0,
+                'price_unit': 200.0,
+                'name': 'Database Administration 25',
+                'product_id': self.product.id,
+                'uom_id': self.product.uom_id.id,
+            })]
+        })
+        res = self.contract._prepare_invoice_data(self.contract)
```

Comment:
Test shouldn't call directly this method, but check result on the invoice after calling `recurring_create_invoice`. Check https://github.com/OCA/contract/pull/73/files#diff-fad070bd2b05e275bf84e1c69de78209L72